sdk/webpubsub/Microsoft.Azure.WebJobs.Extensions.WebPubSub/src/Utilities.cs

```diff
@@ -141,5 +134,83 @@ public static RequestType GetRequestType(WebPubSubEventType eventType, string ev
             }
             return RequestType.Ignored;
         }
+
+        public static bool ValidateSignature(string connectionId, string signature, HashSet<string> accessKeys)
+        {
+            foreach (var accessKey in accessKeys)
+            {
+                var signatures = Utilities.GetSignatureList(signature);
+                if (signatures == null)
+                {
+                    continue;
+                }
+                using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(accessKey));
+                var hashBytes = hmac.ComputeHash(Encoding.UTF8.GetBytes(connectionId));
+                var hash = "sha256=" + BitConverter.ToString(hashBytes).Replace("-", "");
+                if (signatures.Contains(hash, StringComparer.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+                else
+                {
+                    continue;
+                }
+            }
+            return false;
+        }
+
+        public static bool ValidateMediaType(string mediaType, out MessageDataType dataType)
+        {
+            try
+            {
+                dataType = GetDataType(mediaType);
+                return true;
+            }
+            catch (Exception)
+            {
+                dataType = MessageDataType.Binary;
+                return false;
+            }
+        }
+
+        public static bool RespondToServiceAbuseCheck(HttpRequestMessage req, HashSet<string> allowedHosts, out HttpResponseMessage response)
+        {
+            response = new HttpResponseMessage();
+            if (req.Method == HttpMethod.Options || req.Method == HttpMethod.Get)
+            {
+                var requestHosts = req.Headers.GetValues(Constants.Headers.WebHookRequestOrigin);
+                return RespondToServiceAbuseCheck(requestHosts, allowedHosts, out response);
+            }
+            return false;
+        }
+
+        public static bool RespondToServiceAbuseCheck(HttpRequest req, HashSet<string> allowedHosts, out HttpResponseMessage response)
+        {
+            response = new HttpResponseMessage();
+            if (req.Method.Equals("options", StringComparison.OrdinalIgnoreCase) || req.Method.Equals("get", StringComparison.OrdinalIgnoreCase))
+            {
+                req.Headers.TryGetValue(Constants.Headers.WebHookRequestOrigin, out var requestHosts);
+                return RespondToServiceAbuseCheck(requestHosts, allowedHosts, out response);
+            }
+            return false;
+        }
+
+        private static bool RespondToServiceAbuseCheck(IEnumerable<string> requestHosts, HashSet<string> allowedHosts, out HttpResponseMessage response)
+        {
+            response = new HttpResponseMessage();
+            if (requestHosts != null && requestHosts.Any())
+            {
+                foreach (var item in allowedHosts)
+                {
+                    if (requestHosts.Contains(item))
```

Comment:
>requestHosts [](http://example.com/codeflow?start=24&length=12)

use allowedHosts as it is a HashSet? question: shall it case insensitive?

Reply:
item is from allowedHosts, so here compare with requestHosts